account_bank_statement_import_online_transferwise/models/online_bank_statement_provider_transferwise.py

```diff
@@ -0,0 +1,286 @@
+# Copyright 2019 Brainbean Apps (https://brainbeanapps.com)
+# Copyright 2019 Dataplug (https://dataplug.io)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from dateutil.relativedelta import relativedelta
+import dateutil.parser
+from decimal import Decimal
+import itertools
+import json
+import pytz
+import urllib.parse
+import urllib.request
+
+from odoo import models, fields, api, _
+from odoo.exceptions import UserError
+
+
+class OnlineBankStatementProviderTransferwise(models.Model):
+    _inherit = 'online.bank.statement.provider'
+
+    transferwise_api_base = fields.Char(
+        default='https://api.transferwise.com',
+    )
+    transferwise_profile = fields.Char()
+
+    @api.model
+    def values_transferwise_profile(self):
+        api_base = self.env.context.get('api_base')
+        api_key = self.env.context.get('api_key')
+        if not api_key:
+            return []
+        try:
+            url = api_base + '/v1/profiles'
+            data = json.loads(
+                self._transferwise_retrieve(url, api_key),
+                parse_float=Decimal,
+            )
+        except:
+            return []
+        return list(map(
+            lambda entry: (
+                str(entry['id']),
+                '%s %s (personal)' % (
+                    entry['details']['firstName'],
+                    entry['details']['lastName'],
+                )
+                if entry['type'] == 'personal'
+                else entry['details']['name']
+            ),
+            data
+        ))
+
+    @api.model
+    def _get_available_services(self):
+        return super()._get_available_services() + [
+            ('transferwise', 'TransferWise.com'),
+        ]
+
+    @api.multi
+    def _obtain_statement_data(self, date_since, date_until):
+        self.ensure_one()
+        if self.service != 'transferwise':
+            return super()._obtain_statement_data(
+                date_since,
+                date_until,
+            )  # pragma: no cover
+
+        api_base = self.transferwise_api_base
+        api_key = self.password
+        currency = self.currency_id.name
+
+        if date_since.tzinfo:
+            date_since = date_since.astimezone(pytz.utc).replace(tzinfo=None)
+        if date_until.tzinfo:
+            date_until = date_until.astimezone(pytz.utc).replace(tzinfo=None)
+
+        # Get corresponding balance by currency
+        url = api_base + '/v1/borderless-accounts?profileId=%s' % (
+            self.transferwise_profile,
+        )
+        data = json.loads(
+            self._transferwise_retrieve(url, api_key),
+            parse_float=Decimal,
+        )
+        borderless_account = data[0]['id']
+        balance = list(filter(
+            lambda balance: balance['currency'] == currency,
```

Comment:
I'll write down the issues as soon as I find any while debugging.

Here for example, it's possible that `self.currency_id` is not set and variable `currency` is False. In this case, the balance list is always empty.

Reply:
If there's no `currency` for the journal, we're unable to pull statements anyhow, as there's no way to select the proper statement. In what case it can be not set?

Reply:
If you don't enable group "Multi Currencies" for the user you cannot set any currency in the journal. In this case the system takes the company currency I guess.
Instead if you need the currency in the journal always set, I would say to raise an exception.

Reply:
In case it's not set, should I take the company currency?

Reply:
Yes please

Reply:
Done